sound/soc/sof/intel/hda-dsp.c

```diff
@@ -887,10 +887,64 @@ int hda_dsp_suspend(struct snd_sof_dev *sdev, u32 target_state)
 	return snd_sof_dsp_set_power_state(sdev, &target_dsp_state);
 }
 
+static unsigned int hda_dsp_check_for_dma_streams(struct snd_sof_dev *sdev)
+{
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	struct hdac_stream *s;
+	int active_streams;
+	int sd_offset;
+	u32 val;
+
+	list_for_each_entry(s, &bus->stream_list, list) {
+		sd_offset = SOF_STREAM_SD_OFFSET(s);
+		val = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR,
+				       sd_offset);
+		if (val & SOF_HDA_SD_CTL_DMA_START)
+			active_streams = BIT(s->index);
+	}
+
+	return active_streams;
+}
+
+static int hda_dsp_s5_quirk(struct snd_sof_dev *sdev)
+{
+	int ret;
+
+	ret = hda_dsp_ctrl_link_reset(sdev, false);
+	if (ret < 0) {
+		dev_err(sdev->dev, "failed to exit HDA reset (S5 quirk)\n");
+		return -1;
+	}
+
+	usleep_range(500, 1000);
+
+	ret = hda_dsp_ctrl_link_reset(sdev, true);
+	if (ret < 0) {
+		dev_err(sdev->dev, "failed to enter HDA reset (S5 quirk)\n");
+		return -1;
+	}
+
+	return 0;
+}
+
 int hda_dsp_shutdown(struct snd_sof_dev *sdev)
 {
+	unsigned int active_streams;
+	int ret;
+
+	/* ensure DMA cleanup was successful */
+	active_streams = hda_dsp_check_for_dma_streams(sdev);
+
 	sdev->system_suspend_target = SOF_SUSPEND_S3;
-	return snd_sof_suspend(sdev->dev);
+	ret = snd_sof_suspend(sdev->dev);
+
+	if (active_streams) {
+		dev_warn(sdev->dev, "DSP streams 0x%08X still active at shutdown, trying to recover\n",
+			 active_streams);
```

Comment:
that shows you really wanted an |= above if you want to show all active streams.